Add mesh tutorials to the main tutorials branch (#1908)
ansys/pydpf-core: doc/source/user_guide/tutorials/mesh/split_mesh.rst

.. _ref_tutorials_split_mesh:

============
Split a mesh
============

:bdg-mapdl:`MAPDL` :bdg-lsdyna:`LSDYNA` :bdg-fluent:`Fluent` :bdg-cfx:`CFX`

.. include:: ../../../links_and_refs.rst

.. |MeshesContainer| replace:: :class:`MeshesContainer <ansys.dpf.core.meshes_container.MeshesContainer>`
.. |split_mesh| replace:: :class:`split_mesh <ansys.dpf.core.operators.mesh.split_mesh.split_mesh>`
.. |split_on_property_type| replace:: :class:`split_on_property_type <ansys.dpf.core.operators.scoping.split_on_property_type.split_on_property_type>`
.. |from_scopings| replace:: :class:`from_scopings <ansys.dpf.core.operators.mesh.from_scopings.from_scopings>`
.. |ScopingsContainer| replace:: :class:`ScopingsContainer <ansys.dpf.core.scopings_container.ScopingsContainer>`

> .. |PropertyField| replace:: :class:`PropertyField <ansys.dpf.core.property_field.PropertyField>`

Reviewer: Great job Luisa! You made heavy use of substitution definitions which very nice. 

I observed the following concerning these definitions:
- Some definitions already in links_and_refs.rst are being redefined, for example `|MeshesContainer|` and `|ScopingsContainer|`. If definitions in links_and_refs.rst are changed in the future, the changes will not reflect in your examples because local definitions will take precedence.
- You have same substitution definitions in more than one .rst file, for example `|meshes_provider|` common to _get_mesh_from_result_file.rst_ and _extract_mesh_in_split_parts.rst_. If there is a restructuring of the dpf.core package in the future affecting the location `|meshes_provider|` is pointing to, then each of the definitions in your examples have to be located and changed. It is a similar problem to that above.

The solution to this is to move all your substitution definitions to links_and_refs.rst. Then you have only one source of truth. Let me know what you think about the suggested approach.

> This tutorial shows how to split a mesh on a give property.

Reviewer 2: ```suggestion
This tutorial shows how to split a mesh on a given property.
```

            # Import the examples module
            from ansys.dpf.core import examples
            # Import the operators module
            from ansys.dpf.core import operators as ops

            # Define the result file path
            result_file_path_3 = examples.download_fluent_axial_comp()["flprj"]
            # Create the model
            model_3 = dpf.Model(data_sources=result_file_path_3)
            # Get the mesh
            meshed_region_3 = model_3.metadata.meshed_region

    .. tab-item:: CFX

        .. jupyter-execute::

            # Import the ``ansys.dpf.core`` module
            from ansys.dpf import core as dpf
            # Import the examples module
            from ansys.dpf.core import examples
            # Import the operators module
            from ansys.dpf.core import operators as ops

            # Define the result file path
            result_file_path_4 = examples.download_cfx_mixing_elbow()
            # Create the model
            model_4 = dpf.Model(data_sources=result_file_path_4)
            # Get the mesh
            meshed_region_4 = model_4.metadata.meshed_region

.. _ref_first_approach_split_mesh:

First approach
--------------

> This approach consist of splitting an already existing |MeshedRegion| based on a given property. To accomplish

Reviewer 3: ```suggestion
This approach consists in splitting an already existing |MeshedRegion| based on a given property. To accomplish
```

> that goal, you must use the |split_mesh| operator. Currently you can split a mesh by material or eltype.

Reviewer 2: I don't think we are restricted to these properties, are we?

> that goal, you must use the |split_mesh| operator. Currently you can split a mesh by material or eltype.

Reviewer 3: ```suggestion
that end, you must use the |split_mesh| operator. Currently you can split a mesh by material or eltype.
```

The split mesh parts are stored in the DPF collection called |MeshesContainer|, where they are ordered by *labels*.
When you use the |split_mesh| operator, each split mesh part has two different *labels*:

> - A "body" *label*;
> - A *label* with the property used to split the mesh.

Reviewer 3: ```suggestion
- A "body" *label*
- A *label* with the property used to split the mesh
```

Here, we split the |MeshedRegion| by material.

.. tab-set::

    .. tab-item:: MAPDL

        .. jupyter-execute::

            # Split the mesh by material

>             meshes_11 = ops.mesh.split_mesh(mesh=meshed_region_1,property="mat").eval()

Reviewer 3: ```suggestion
            meshes_11 = ops.mesh.split_mesh(mesh=meshed_region_1, property="mat").eval()
```

            # Print the meshes
            print(meshes_11)

    .. tab-item:: LSDYNA

        .. jupyter-execute::

            # Split the mesh by material

>             meshes_21 = ops.mesh.split_mesh(mesh=meshed_region_2,property="mat").eval()

Reviewer 2: Their are some formatting errors :)
```suggestion
            meshes_21 = ops.mesh.split_mesh(mesh=meshed_region_2, property="mat").eval()
```

            # Print the meshes
            print(meshes_21)

    .. tab-item:: Fluent

        .. jupyter-execute::

            # Split the mesh by material

>             meshes_31 = ops.mesh.split_mesh(mesh=meshed_region_3,property="mat").eval()

Reviewer 3: ```suggestion
            meshes_31 = ops.mesh.split_mesh(mesh=meshed_region_3, property="mat").eval()
```

            # Print the meshes
            print(meshes_31)

    .. tab-item:: CFX

        .. jupyter-execute::

            # Split the mesh by material

>             meshes_41 = ops.mesh.split_mesh(mesh=meshed_region_4,property="mat").eval()

Reviewer 3: ```suggestion
            meshes_41 = ops.mesh.split_mesh(mesh=meshed_region_4, property="mat").eval()
```

            # Print the meshes
            print(meshes_41)

.. _ref_second_approach_split_mesh:

Second approach
---------------

> This approach consists of splitting the |Scoping| of a given |MeshedRegion| based on a given property and then creating

Reviewer 3: ```suggestion
This approach consists in splitting the |Scoping| of a given |MeshedRegion| based on a given property and then creating
```

a new |MeshedRegion| for each split |Scoping|.

To accomplish this goal you must follow these steps:

#. Use the |split_on_property_type| operator to split the mesh |Scoping|.
    This operator splits a |Scoping| on a given property (elshape and/or material, since 2025R1 it supports any
    scalar property field name contained in the mesh property fields). The split |Scoping| is stored in the DPF
    collection called |ScopingsContainer|, where they are ordered by *labels*. In this case, you get *labels* with
    the property used to split the |Scoping|.

#. Create the split |MeshedRegion| objects using the |from_scopings| operator for the |Scoping| of interest.
    The split parts are stored in the DPF collection called |MeshesContainer| where they are also ordered by *labels*.
    These *labels* are corresponding to the "mat" labels gotten with the |split_on_property_type| operator.

Here, we split the mesh scoping by material and create a |MeshedRegion| for all the split |Scoping| in the
|ScopingsContainer|.

.. tab-set::

    .. tab-item:: MAPDL

        .. jupyter-execute::

>             # Define the scoping split by material
>             split_scoping_1 = ops.scoping.split_on_property_type(mesh=meshed_region_1, label1="mat").eval()
>             # Get the split meshes
>             meshes_12 = ops.mesh.from_scopings(scopings_container=split_scoping_1,mesh=meshed_region_1).eval()
>             # Print the meshes
>             print(meshes_12)
> 
>     .. tab-item:: LSDYNA
> 
>         .. jupyter-execute::
> 
>             # Define the scoping split by material
>             split_scoping_2 = ops.scoping.split_on_property_type(mesh=meshed_region_2, label1="mat").eval()
>             # Get the split meshes
>             meshes_22 = ops.mesh.from_scopings(scopings_container=split_scoping_2,mesh=meshed_region_2).eval()
>             # Print the meshes
>             print(meshes_22)
> 
>     .. tab-item:: Fluent
> 
>         .. jupyter-execute::
> 
>             # Define the scoping split by material
>             split_scoping_3 = ops.scoping.split_on_property_type(mesh=meshed_region_3, label1="mat").eval()
>             # Get the split meshes
>             meshes_32 = ops.mesh.from_scopings(scopings_container=split_scoping_3,mesh=meshed_region_3).eval()
>             # Print the meshes
>             print(meshes_32)
> 
>     .. tab-item:: CFX
> 
>         .. jupyter-execute::
> 
>             # Define the scoping split by material
>             split_scoping_4 = ops.scoping.split_on_property_type(mesh=meshed_region_4, label1="mat").eval()
>             # Get the split meshes
>             meshes_42 = ops.mesh.from_scopings(scopings_container=split_scoping_4,mesh=meshed_region_4).eval()
>             # Print the meshes
>             print(meshes_42)

Reviewer 3: ```suggestion
            # Define the scoping split by material
            split_scoping_1 = ops.scoping.split_on_property_type(mesh=meshed_region_1, label1="mat").eval()
            # Get the split meshes
            meshes_12 = ops.mesh.from_scopings(scopings_container=split_scoping_1, mesh=meshed_region_1).eval()
            # Print the meshes
            print(meshes_12)

    .. tab-item:: LSDYNA

        .. jupyter-execute::

            # Define the scoping split by material
            split_scoping_2 = ops.scoping.split_on_property_type(mesh=meshed_region_2, label1="mat").eval()
            # Get the split meshes
            meshes_22 = ops.mesh.from_scopings(scopings_container=split_scoping_2, mesh=meshed_region_2).eval()
            # Print the meshes
            print(meshes_22)

    .. tab-item:: Fluent

        .. jupyter-execute::

            # Define the scoping split by material
            split_scoping_3 = ops.scoping.split_on_property_type(mesh=meshed_region_3, label1="mat").eval()
            # Get the split meshes
            meshes_32 = ops.mesh.from_scopings(scopings_container=split_scoping_3, mesh=meshed_region_3).eval()
            # Print the meshes
            print(meshes_32)

    .. tab-item:: CFX

        .. jupyter-execute::

            # Define the scoping split by material
            split_scoping_4 = ops.scoping.split_on_property_type(mesh=meshed_region_4, label1="mat").eval()
            # Get the split meshes
            meshes_42 = ops.mesh.from_scopings(scopings_container=split_scoping_4, mesh=meshed_region_4).eval()
            # Print the meshes
            print(meshes_42)
```
